driver/src/orderbook/streamed/order.rs

```diff
@@ -0,0 +1,125 @@
+use super::*;
+use crate::contracts::stablex_auction_element;
+use crate::models::Order as ModelOrder;
+use error::Error;
+use serde::{Deserialize, Serialize};
+
+#[derive(Copy, Clone, Debug, Deserialize, Serialize)]
+pub struct Order {
+    pub buy_token: TokenId,
+    pub sell_token: TokenId,
+    pub valid_until: BatchId,
+    pub valid_from: BatchId,
+    price_numerator: u128,
+    price_denominator: u128,
+    // Invariant: used_amount + pending_used_amount.1 <= price_denominator
+    used_amount: u128,
+    pending_used_amount: (BatchId, u128),
+}
+
+impl Order {
+    pub fn new(
+        buy_token: TokenId,
+        sell_token: TokenId,
+        valid_from: BatchId,
+        valid_until: BatchId,
+        price_numerator: u128,
+        price_denominator: u128,
+    ) -> Self {
+        Self {
+            buy_token,
+            sell_token,
+            valid_from,
+            valid_until,
+            price_numerator,
+            price_denominator,
+            used_amount: 0,
+            pending_used_amount: (0, 0),
+        }
+    }
+
+    pub fn is_valid_in_batch(&self, batch_id: BatchId) -> bool {
+        self.valid_from <= batch_id && batch_id <= self.valid_until
+    }
+
+    pub fn trade(&mut self, used_amount: u128, batch_id: BatchId) -> Result<(), Error> {
+        if !self.has_limited_amount() {
+            return Ok(());
+        }
+        let new_used_amount = self.get_used_amount(batch_id);
+        let new_pending_used_amount = if self.pending_used_amount.0 == batch_id {
+            self.pending_used_amount
+                .1
+                .checked_add(used_amount)
+                .ok_or(Error::MathOverflow)?
+        } else {
+            used_amount
+        };
+        // Make sure the invariant still holds.
+        if new_pending_used_amount > self.price_denominator {
+            return Err(Error::TradeByMoreThanOrderLimit(self.price_denominator));
+        }
+        self.pending_used_amount = (batch_id, new_pending_used_amount);
+        self.used_amount = new_used_amount;
```

Comment:
Why is this written? I don't see it being changed in this method. After reviewing get_used_amount carefully I can see that it takes the pending_used_amount into account which in this method might get overridden and thus we have to write this value otherwise we lose this information.

This logic is extremely subtle and hidden in this code. Instead I'd suggest we turn `get_used_amount ` into `update_used_amount`, which takes care of applying and resetting the pending_used_amount if necessary. This way we can also get rid of the if/else statement when computing the new_pending_amount as if it was reset it would be 0 so we can always add `used_amount` to it.